### PR TITLE
Enable tracing and script autograd tests

### PR DESCRIPTION
**Pull request by zdevito**
This commit adds additional pathways to the autograd tests to test
when the methods are run inside of a trace, or directly written using
script. These tests have uncovered many bugs and limited functionality
in the trace/script pathway, and these failing parts of the tests
are disabled using new exclusion sets. The size of these sets will shrink
rapidly as the bugs are fixed, but are currently quite large.

Rather than use gradcheck, these tests just compare trace/script gradients against the autograd. This approach is faster, so the test_autograd tests do not run appreciably slower with them enabled.

### CODE REVIEW

**Comment by apaszke** [comment]
Can we please move those to `test_jit.py`? You should be able to safely import `test_autograd` from there, but I'd rather avoid dumping extra 400 lines of code that have no connection to autograd in `test_autograd`

File: test/test_autograd.py [hunk 1]
```diff
@@ -3189,17 +3423,127 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
...
+        l1_test, recording_tensors, create_graph=True, allow_unused=allow_unused)
+    l2_test = (allSum(grads_test) * l1_test)
+    grads2_test = torch.autograd.grad(l2_test, recording_tensors, allow_unused=allow_unused)
+
+    self.assertEqual(outputs, outputs_test)
+    self.assertEqual(grads, grads_test)
+    for g2, g2_test in zip(grads2, grads2_test):
+        if g2 is None and g2_ge is None:
+            continue
+        self.assertTrue(torch.allclose(g2, g2_test, atol=5e-4, rtol=1e-4))
```
Comment (apaszke) [inline test/test_autograd.py]: We already have a helper like this in `TestJit` (`checkTrace`), which is an additional reason to move it there.

**Comment by zdevito** [comment]
I can move all the helper functions to a separate file, but moving the 4 calls to check_against_reference  outside of the test harness will require serious refactoring or code duplication of the thing that generates the reference tests, which I don't think buys us much except a sense of cleanliness.

File: test/test_autograd.py (hunk 1 above)
Comment (zdevito) [reply test/test_autograd.py]: I can factor this into a common function

File: test/test_autograd.py [hunk 2]
```diff
@@ -3104,6 +3108,236 @@ def unpack_variables(args):
...
+    'test_unsqueeze_last_neg0',
+    'test_unsqueeze_middle_neg0',
+    'test_split_dim',
+    'test_split_dim_neg0',
+    'test_gesv',
+    'test_inverse',
+}
+
+# known to be failing in script
+EXCLUDE_SCRIPT = {
```
Comment (ezyang) [inline test/test_autograd.py]: Wouldn't it be better to run these tests, and then fail if they unexpectedly pass? Will help you remember to re-enable tests when they start passing.

**Comment by jamesr66a** [comment]
@zdevito  can we get logs of all the failures for the excluded operators so we can start triaging?

**Comment by apaszke** [comment]
I still think we should put this in `test_jit.py`. Otherwise running `test_autograd` checks many components that are completely irrelevant in this case. What dependencies would that have? You can just import the specs from that file. The format doesn't really change, and the code that parses it into arguments is effectively 15 lines of code. The rest is the autograd test template, which we don't need.

**Review by apaszke — CHANGES_REQUESTED** [review]
See above

**Comment by zdevito** [comment]
Current failures: https://gist.github.com/zdevito/6d62435b68edf027a98ed6c9d96ef3df

**Comment by zdevito** [comment]
I moved the tests to test_jit. @jamesr66a 's https://github.com/pytorch/pytorch/pull/8451 also includes changes to checkTrace, so I haven't factored it out here. I want to just merge these separately than do the refactor in a follow up PR.

**Review by apaszke — APPROVED** [review]

**Pull request merged**